Extend use of EXTERNAL_POSITION_ESTIMATE data

priseborough opened the pull request:
Optical navigation technology can provide data in global coordinates, but can also have large variations in processing delay  between images which is not catered for by either of the interfaces listed above. Optical navigation can either run on a GCS based system using transmitted camera data or on an onboard companion computer.

The GCS processing case was previously supported in a limited way by the MAV_CMD_EXTERNAL_POSITION_ESTIMATE message. It was developed originally to support infrequent operator position fixes from a GCS with up to 5 seconds latency.  These data were used only to reset the position inside the EKF and is not used as observations for state updates. This means that for vehicles doing wind relative dead reckoning, more regular updates could not be used to correct wind and vehicle velocity states for improved dead reckoning accuracy. 

Onboard high rate position and attitude data with small and consistent  latencies are currently supported via the VISION_POSITION_ESTIMATE, GLOBAL_VISION_POSITION_ESTIMATE and VICON_POSITION_ESTIMATE messages which the EKF processes using the writeExtNavData interface. This is limited to local coordinates and cannot handle large varying processing delays.

This pull request enables the use of global position estimates from either GCS or onboard companion computers. Latencies of up to 5 seconds can be handled and the processing of these estimates as EKF observations means that vehicle and wind velocity states can be corrected.  It also means that innovation consistency checks will be performed which increases protection from error spikes.

GCS based data are supported by extension of the EXTERNAL_POSITION_ESTIMATE command message handling to enable its use with higher rate messages.

The same functionality is also supported for on-board companion computer processing via the GLOBAL_POSITION_SENSOR streaming message.

The EK3_OPTIONS parameter is used to enable and configure the feature. Setting bit 3 enables the feature and MAV_CMD_EXTERNAL_POSITION_ESTIMATE data will be treated as a measurement. This bit is off by default. Setting bit 4 causes the offset between the MAV_CMD_EXTERNAL_POSITION_ESTIMATE to be learned when GPS quality is sufficient to pass the same checks used for EKF alignment. This offset is then applied to the data when required for navigation. This option is useful if the MAV_CMD_EXTERNAL_POSITION_ESTIMATE message is derived from an odometry technique that drifts over time.

If the EKF_OPTIONS bit 0 is set (jamming expected option), then if the EKF GPS good to align checks fail and MAV_CMD_EXTERNAL_POSITION_ESTIMATE data is available, GPS use will be discontinued and MAV_CMD_EXTERNAL_POSITION_ESTIMATE data used instead until GPS good to align checks pass. The GPS good to align checks are controlled by the EK3_GPS_CHECK and EK3_CHECK_SCALE parameters.

A basic autotest has been included for the MAV_CMD_EXTERNAL_POSITION_ESTIMATE message. 

The following figures are from additional SITL testing using an offboard Gazebo derived vision nav sim using the MAV_CMD_EXTERNAL_POSITION_ESTIMATE message.

The following plots show the Copter vehicle type test with GPS turned off halfway through the run.

<img width="603" height="576" alt="Screenshot 2025-12-17 at 11 50 56 AM" src="https://github.com/user-attachments/assets/108e09ca-fd4e-4925-8c48-3b3b606a692a" />

The position innovations get noisy when the GPS stops as the EKF starts using the MAV_CMD_EXTERNAL_POSITION_ESTIMATE data.

<img width="1023" height="538" alt="Screenshot 2025-12-17 at 11 46 14 AM" src="https://github.com/user-attachments/assets/2377d533-ad77-49a2-8942-9e8baab4e1ef" />

GPS velocity innovations show when GPS data stops.

<img width="1045" height="541" alt="Screenshot 2026-01-20 at 8 32 37 AM" src="https://github.com/user-attachments/assets/659ae5b5-fcdb-4b57-a7c9-3efe45162847" />

The simulated measurements have significant lag and jitter which are handled by the EKF

<img width="1036" height="544" alt="Screenshot 2025-12-17 at 11 58 15 AM" src="https://github.com/user-attachments/assets/1162e204-f52b-4597-a27d-7ace9b6c06f7" />

A SITL test was performed using a fixed wing model

<img width="602" height="574" alt="Screenshot 2025-12-17 at 12 03 04 PM" src="https://github.com/user-attachments/assets/cfb133ac-bc1b-4ea1-b644-2224dbc0f8a7" />

Measurements had high jitter and latency

<img width="1079" height="543" alt="Screenshot 2025-12-17 at 12 04 52 PM" src="https://github.com/user-attachments/assets/72306256-0c8f-4c63-a790-7b7c36832661" />

The EKF handles the larger position fix errors in this application - the fallover from GPS to using vision nav is obvious

<img width="1033" height="535" alt="Screenshot 2025-12-17 at 12 05 33 PM" src="https://github.com/user-attachments/assets/7e919b9c-f07c-4a95-8d5b-c8a619938c2d" />


The following figures are from additional SITL testing using an offboard Gazebo derived vision nav sim using the GLOBAL_POSITION_SENSOR message.

<img width="603" height="578" alt="Screenshot 2026-06-09 at 7 03 43 PM" src="https://github.com/user-attachments/assets/8e53ba8e-5f30-4967-9ee7-a7e5d72847a6" />

GPS is disabled
<img width="1053" height="542" alt="Screenshot 2026-06-09 at 7 10 05 PM" src="https://github.com/user-attachments/assets/8f068b83-e72a-4178-ba68-7d687fb2debb" />

Position innovations from the optical navigation position estimates are noisier, but zero mean.
<img width="1018" height="526" alt="Screenshot 2026-06-09 at 7 09 36 PM" src="https://github.com/user-attachments/assets/25f78f37-0de9-4d3d-be97-593faebc5816" />


The EKF tracks the position fixes
<img width="1061" height="545" alt="Screenshot 2026-06-09 at 7 16 48 PM" src="https://github.com/user-attachments/assets/47f5a28a-8914-4fdd-86c0-956daa6703bf" />

A position standard deviation of 11.5m was set by the optical nav
<img width="1030" height="534" alt="Screenshot 2026-06-09 at 7 17 45 PM" src="https://github.com/user-attachments/assets/3c375e63-6957-42a3-8ed6-4d954042c708" />

Simulared measurements have up to 1150msec of  latency and are handled by the EKF.
<img width="1068" height="546" alt="Screenshot 2026-06-09 at 7 20 51 PM" src="https://github.com/user-attachments/assets/017a2ef1-d5f2-4f73-8c0d-1e589f7bfe4b" />

priseborough committed:
autotest: Add test for use of regular EXTERNAL_POSITION_ESTIMATE data

priseborough committed:
AP_NavEKF3: Add option to use setLatLng interface as a fusion data source

AP_NavEKF3: Fix setLatLng GPS handover logic

AP_NavEKF3: Ensure GPS is declared bad if fix status less than 3D

AP_NavEKF3: fix comment typo

AP_NavEKF3: Allow a longer gap between setLatLng before unblocking bad GPS

AP_NavEKF3: Fix public vs internal origin bug affecting setLatLng offset

AP_NavEKF3: Make learning of setLatLng offset a parameter option

AP_NavEKF3: Reset to GPS if prior setLatLng data suspect

AP_NavEKF3: Rework switch between setlatLng and GPS

AP_NavEKF3: Don't use stale data

AP_NavEKF3: Improve robustness to manouevre induced prediction errors

AP_NavEKF3: Provide dal.log_SetLatLng with data received time stamp

AP_NavEKF3: Update documentation for EK3_OPTIONS parameter

AP_NavEKF3: Fix comment typo

AP_NavEKF3: Don't change GPS re-aquisition behaviour unless option set

AP_NavEKF3: #define option to save flash on small boards

AP_NavEKF3: Fix array index typo and document

AP_NavEKF3: Remove unwanted wind covariance reset

AP_NavEKF3: Use existing mechanism to clear gpsGoodToAlign

Achieves the same result as before provided the jamming expected option is set.

AP_NavEKF3: handle GPS receivers that don't report accuracy

AP_NavEKF3: Fix documentation typo

AP_NavEKF3: #define fix for small boards build error

AP_NavEKF3: Prevent simultaneous use of viso and setLatLng data

Also update parameter description to include this information.

AP_NavEKF3: Update documentation for EK3_OPTIONS

AP_NavEKF3: Fix build error from rebase

priseborough committed:
AP_DAL: Add received time stamp to RSLL data for diagnostics

The difference between receive and measurement time is important for debugging this interface.

AP_DAL: Add documentation for RSLL

Co-authored-by: Peter Barker <pb-gh@barker.dropbear.id.au>

AP_DAL: fix typo

Co-authored-by: Peter Barker <pb-gh@barker.dropbear.id.au>

AP_DAL: Fix timestamp descriptions for RSLL log data

AP_DAL: Fix meta data in RSLL message

AP_DAL: fix documentation typo

peterbarker commented on libraries/AP_DAL/LogStructure.h:
```diff
@@ -517,7 +517,8 @@ struct log_RSLL {
     float posAccSD; // horizontal position 1 STD uncertainty (m)
-    uint32_t timestamp_ms;
+    uint32_t timestamp_ms; // time the measurement was taken (msec)
+    uint32_t received_ms; // tine the neasurement was received by the EKF (msec)
```
```suggestion
    uint32_t received_ms; // tine the measurement was received by the EKF (msec)
```

peterbarker commented on libraries/AP_DAL/LogStructure.h:
```diff
@@ -649,7 +650,7 @@ struct log_RTER {
       "REPH", "fffffffffIIH", "PX,PY,PZ,Q1,Q2,Q3,Q4,PEr,AEr,TS,RT,D", "------------", "------------" }, \
     { LOG_RSLL_MSG, RLOG_SIZE(RSLL),                         \
-      "RSLL", "IIfI", "Lat,Lng,PosAccSD,TS", "DU--", "GG--" }, \
+      "RSLL", "IIfII", "Lat,Lng,PosAccSD,MTS,RTS", "DU---", "GG---" }, \
```
```suggestion
      "RSLL", "IIfII", "Lat,Lng,PosAccSD,MTS,RTS", "DU-ss", "GG-CC" }, \
```

(mark fields as time in ms)

peterbarker commented on libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp:
```diff
@@ -570,6 +570,8 @@ void NavEKF3_core::readGpsData()
     if (gps.status(selected_gps) < AP_DAL_GPS::GPS_OK_FIX_3D) {
         // report GPS fix status
         gpsCheckStatus.bad_fix = true;
+        gpsGoodToAlign = false;
```
Should we gate this behind the option bit to keep this NFC if you don't have the new feature enabled?

peterbarker commented on libraries/AP_NavEKF3/AP_NavEKF3_core.h:
```diff
@@ -150,8 +150,9 @@ class NavEKF3_core : public NavEKF_core_common
 
     // Write the last calculated NE position relative to the reference point (m).
     // If a calculated solution is not available, use the best available data and return false
+    // set delayd to true if solution required at fusion time horizon
```
```suggestion
    // set delayed to true if solution required at fusion time horizon
```

peterbarker commented on libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp:
```diff
@@ -190,35 +190,92 @@ void NavEKF3_core::ResetPosition(resetDataSource posResetSource)
 {
-    if ((imuSampleTime_ms - lastGpsPosPassTime_ms) < frontend->deadReckonDeclare_ms ||
+    if (((imuSampleTime_ms - lastGpsPosPassTime_ms) < frontend->deadReckonDeclare_ms && !frontend->option_is_enabled(NavEKF3::Option::SetLatLngFusion))  ||
         (PV_AidingMode == AID_NONE)
```
So you can't move from "No idea where I am" using this mechanism?

peterbarker reviewed (changes requested):
This obviously does change behaviour for users that don't have the option enabled.  Would be nice to have those spelled out.

Considering that the *vast* majority of our users won't be using this feature, I think we really do need `EK3_FEATURE_EXTERNAL_POSITION_FUSION`
```
-----------  -----  ----------  ------  ----  -----  -----  ---
Board        blimp  bootloader  copter  heli  plane  rover  sub
Pixhawk1-1M  440    *           440     448   440    448    440
-----------  -----  ----------  ------  ----  -----  -----  ---
```

CI must pass; most of the errors will be fixed by adding documentation for your changed/added log message fields.

snktshrma commented on libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp:
```diff
@@ -190,35 +190,92 @@ void NavEKF3_core::ResetPosition(resetDataSource posResetSource)
-    ResetPositionNE(newPosNE.x,newPosNE.y);
+                ResetPositionNE(velPosObs[3], velPosObs[4]);
+
+                zeroRows(P,22,23);
```
Is the zeroing of wind covariances intentional here? I didn't see this in other position reset code paths either and wanted to understand the reasoning.

priseborough replied:
Yes. Have removed it.

priseborough committed:
autotest: use do_RTL

Co-authored-by: Peter Barker <pb-gh@barker.dropbear.id.au>

priseborough commented on libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp:
```diff
@@ -680,6 +682,22 @@ void NavEKF3_core::readGpsData()
+            if (frontend->option_is_enabled(NavEKF3::Option::JammingExpected)) {
+                useSetLatLngAsMeasurement = !gpsGoodToAlign;
+            } else {
+                useSetLatLngAsMeasurement = gpsPosAccuracy > safe_sqrt(P[7][7]+P[9][8]);
```
Fixed

rishabsingh3003 commented on libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp:
```diff
@@ -680,6 +686,25 @@ void NavEKF3_core::readGpsData()
+                useSetLatLngAsMeasurement = !gpsGoodToAlign;
+            } else {
+                // GPS is reporting worse accuracy than the current horizontal position estimate
+                useSetLatLngAsMeasurement = gpsPosAccuracy > safe_sqrt(P[7][7]+P[8][8]);
```
This will always be false for GPS's that dont report accuracy. Probably needs a check for positive values

priseborough replied:
I've done that and added comments to make the intent clearer

priseborough committed:
GCS_MAVLink: Add use of GLOBAL_POSITION_SENSOR

GCS_MAVLink: Add missing micro to milli seconds processing time conversion

GCS_MAVLink: Add case for MAVLINK_MSG_ID_GLOBAL_POSITION_SENSOR

priseborough commented:
I've rebased on latest master (no conflicts) and have added support for the GLOBAL_POSITION_SENSOR message. I'll follow up with some test logs from the sponsor.

rishabsingh3003 commented on libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp:
The cov are being reset unconditionally? Is that correct when the new bits in EK3_OPTIONS are set?

priseborough replied:
Looks like that line and the setting of P[7][7] and P[8][8] were re-introduced by mistake during the last rebase. I have removed them and moved other new lines only required by the new functionality into the EK3_FEATURE_EXTERNAL_POSITION_FUSION  code section.

priseborough committed:
AP_NavEKF3: Remove lines that were restored by mistake during rebase

priseborough committed:
AP_NavEKF3: Clean up after rebase

priseborough committed:
GCS_MAVLink: Fix build error with small boards

handle_global_position_sensor can't be used with small boards

priseborough committed:
mavlink: update to latest master to include GLOBAL_POSITION_SENSOR

priseborough committed:
GCS_MAVLink: Don't nest inside #if HAL_VISUALODOM_ENABLED

rishabsingh3003 commented on libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp:
```diff
@@ -1090,7 +1118,12 @@ void NavEKF3_core::writeExtNavData(const Vector3f &pos, const Quaternion &quat,
+    // data sources.
+    if (((timeStamp_ms - extNavMeasTime_ms) < frontend->extNavIntervalMin_ms) ||
+        !statesInitialised ||
+        lastSetlatLngPassTime_ms - imuSampleTime_ms < 5000) {
```
Isn't `lastSetlatLngPassTime_ms - imuSampleTime_ms` always less than 5000?

priseborough replied:
Yes, thats the idea. The writeExtNavData  interface with local position external nav measurements should not be used at the same time as setLatLng that uses location data as they will fight.

rishabsingh3003 commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -4172,6 +4172,34 @@ void GCS_MAVLINK::handle_vision_speed_estimate(const mavlink_message_t &msg)
+    timestamp_ms -= m.processing_time / 1000;
+
+    AP_AHRS &ahrs = AP::ahrs();
+    ahrs.handle_external_position_estimate(loc, accuracy, timestamp_ms);
```
So every time this mavlink message is sent, we run a full Kalman update from this thread? Does this not need a buffer and we then fuse sequentially like other measurements?

priseborough replied:
The thinking behind doing it this way was:

1) The potential large time delay put the measurement outside the normal fusion time horizon delay window so the measurement was projected to the current EKF fusion time horizon which avoided the cost of a new buffer for each core.
2) Because it is a direct state observation of only 2 states, the fusion step for each core is fairly cheap.
3) It was an extension of the exisiting function used to apply the data as a reset.


I will add protection inside NavEKF3_core::setLatLng to prevent it being called faster than 5Hz

priseborough replied:
Protection against calling faster than 5Hz has been added

rishabsingh3003 commented on libraries/GCS_MAVLink/GCS_Common.cpp:
```diff
@@ -4172,6 +4172,34 @@ void GCS_MAVLINK::handle_vision_speed_estimate(const mavlink_message_t &msg)
+{
+    mavlink_global_position_sensor_t m;
+    mavlink_msg_global_position_sensor_decode(&msg, &m);
+    if (m.flags == GLOBAL_POSITION_FLAGS::GLOBAL_POSITION_UNHEALTHY) {
```
Is this not a bitmask?

priseborough replied:
I'll check

priseborough replied:
Yes it is. I've pushed a fix.

rishabsingh3003 commented on libraries/AP_NavEKF3/AP_NavEKF3_core.h:
```diff
@@ -1672,6 +1673,16 @@ class NavEKF3_core : public NavEKF_core_common
+
+#if EK3_FEATURE_EXTERNAL_POSITION_FUSION
+    // fusion of data received via the setLatLng interface
+    Vector2F setLatLngPosOffsetNE; // NE position offset in metres added to the received position
```
Do we need some of these initialzied in `InitialiseVariable`?

priseborough replied:
Probably the offset and the time stamp(s) to be sure. I'll add them

priseborough replied:
I've added all the class variables for this feature to the reset function to be safe.

rishabsingh3003 commented on libraries/AP_NavEKF3/AP_NavEKF3_PosVelFusion.cpp:
```diff
@@ -627,6 +685,15 @@ void NavEKF3_core::SelectVelPosFusion()
     selectHeightForFusion();
 
+#if EK3_FEATURE_EXTERNAL_POSITION_FUSION
+    if (gpsDataToFuse && gpsGoodToAlign && ((imuSampleTime_ms - lastResetlatLngTime_ms) < 10000 || filterStatus.flags.dead_reckoning)) {
```
Does this need a check for `lastResetlatLngTime_ms` to be non-zero? Otherwise it will always be true on boot?

priseborough replied:
I'll add the check.

priseborough replied:
In theory, yes if a plane was launched without a compass within 10 seconds of boot, this could force gps checks to be declared as passed and trigger this condition. I've added a check.

rishabsingh3003 commented on Tools/autotest/arducopter.py:
```diff
@@ -5090,6 +5090,64 @@ def VisionPosition(self):
+            self.set_parameter("EK2_GPS_TYPE", 3)
+        if ahrs_ekf_type == 3:
+            self.set_parameters({
+                "EK3_OPTIONS": 24, # SetLatLngFusion and SetLatLngOffset option activated
```
Is this the right bitmask? Bit 3 is AGL KF

priseborough replied:
Will check.

priseborough replied:
I've changed it to 16+32 = 48

Gabe-OI commented:
Hello, I'm a computer engineer from Overhead Intelligence. I just wanted to quickly say that I flew this PR (along with PR #28279 and a simple velocity state edit to bridge compatibility between the two) rebased to commit f808f78ce5a of master. I've been experimenting with an external navigation system based on Starlink and MAV_CMD_EXTERNAL_POSITION_ESTIMATE (43003) for months now, and this PR's approach of treating estimates as a fusion rather than a state set has produced unbelievable results. The mean EKF drift for our most recent flight was 7 meters, far closer to truth overall than the estimates being fed to it during pure dead-reckon, and at one point was able to go for over a minute without any estimates while only drifting ~10 meters. 

At a glance, I firmly believe this rework will be essential to stable operations in denied environments, although I'm aware that its use along with an older PR targeting a tangential feature means this wasn't a "pure" test. Within the custom image I compiled for this denied flight, the PR occupies 1548 B of flash... I can provide a copy of the image and the log from the flight that ran it upon request. I'm also open to suggestions on ideal conditions for testing this PR in the future!
